### ISSUE TITLE
Context-menu option to add current page as attachment

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero-connectors#102]
This would restore functionality of the "Attach Snapshot of Current Page" and "Attach Link to Current Page" options in Zotero for Firefox. (Dragging the link or icon works, but it's annoying if you don't use Cmd-Tab.)

https://forums.zotero.org/discussion/comment/269248/#Comment_269248

**Comment by adomasven** [comment]
I am rather worried about overcluttering the button context menu with confusingly similar options. I see the purpose, but I think this would be confusing when we also have "Save to Zotero (Web Page with Snapshot)" and "Save to Zotero (Web Page without Snapshot)". 

We could consider adding this as a keyboard shortcut, although I am not sure how to make that discoverable.

**Comment by adomasven** [comment]
I am rather worried about overcluttering the button context menu with confusingly similar options. I see the purpose, but I think this would be confusing when we also have "Save to Zotero (Web Page with Snapshot)" and "Save to Zotero (Web Page without Snapshot)". 

We could consider adding this as a keyboard shortcut, although I am not sure how to make that discoverable.

**Comment by dstillman** [comment]
I don't think we need to worry about that too much — if you're right-clicking on the button, you're already looking for advanced options, and these are existing options that people use. Though it would be much better if we could add a separate, which I'm guessing we can't?

In any case, this functionality is most important for when an attachment isn't saved by the translator, so we could potentially limit this to PDFs and images, and so only show "Save to Zotero (PDF)", "Attach PDF to Zotero Item", and "Attach Link to PDF to Zotero Item" (and same for "Image", maybe).

But that would mean you couldn't add easily a link or snapshot for a webpage to an existing item. Not sure how common that is. Possible it's rare enough that drag-and-drop (or, in the case of link, copying and pasting the URL) is sufficient.

Not sure if we have the ability to check for whether a regular item is selected in Zotero on right-click and update the menu accordingly, but if so we could not show these options when not applicable.

**Comment by adomasven** [comment]
> Though it would be much better if we could add a separator or submenu, which I'm guessing we can't?

We can do both separators and submenus. Except maybe safari, where the menu is actually a popover, so submenus would require some sort of creative solution. But separators should suffice for now anyway.

> Not sure if we have the ability to check for whether a regular item is selected in Zotero on right-click and update the menu accordingly, but if so we could not show these options when not applicable.

We might want to update current `/connector/ping` to respond with currently selected collection and item and just store the last one. Not an ideal solution, but better than naught. Really one more place where webSockets would really come in handy.

> In any case, this functionality is most important for when an attachment isn't saved by the translator, so we could potentially limit this to PDFs and images, and so only show "Save to Zotero (PDF)", "Attach PDF to Zotero Item", and "Attach Link to PDF to Zotero Item" (and same for "Image", maybe).

Possibly more confusing/annoying for users than just having it show up everywhere. I'd say it's better to display/allow this action on every page. I still do think that we should consider keyboard shortcuts. They are probably pretty weird/technical for an average user, but this is intended to be semi-advanced functionality anyway.

**Comment by dstillman** [comment]
This isn't really that advanced (and if you're advanced, dragging-and-dropping the page proxy icon through Cmd-Tab is trivial — it's people who don't know how to do that who need this). I'm pretty skeptical of using any other shortcuts in the browser. (All shortcuts other than saving in Zotero for Firefox were restricted to the Zotero pane.) But agreed about showing for everything, particularly if we can do separator/submenu.

**Comment by dstillman** [comment]
> No way to check on right-click, the event handler is synchronous.

But not possible to kick off an async process and update the menu after it's open? (Not ideal, but we do that now for the context menus in Zotero itself.)

**Comment by adomasven** [comment]
> But not possible to kick off an async process and update the menu after it's open? (Not ideal, but we do that now for the context menus in Zotero itself.)

Certainly for Safari, possibly for BrowserExt. Wouldn't be surprised if it worked on FX but not on Chrome or similar.

**Comment by adomasven** [comment]
I'm going to go through with this. We'll only show the context-menu option when Zotero is online. We'll add an endpoint to the server `/connector/attachSnapshot` which will return the currently selected item info (or `500 {libraryEditable: false}`) and we'll display that in the progress window.

**Comment by dstillman** [comment]
We should use SSE and only show when an appropriate item is selected. It shouldn't show at all when a note or attachment is selected.

(With SSE we might even want to consider showing at least the beginnings of collection/item names in the menu options, to give users confidence that actions will save to the right place.)

**Comment by dstillman** [comment]
Once we get this in, it [would be nice](https://forums.zotero.org/discussion/comment/301443/#Comment_301443) to 1) use a clearer label for PDFs and 2) include the name of the selected item. So, e.g., `Attach PDF to “Psychological testing”`. (Not sure if we're sending down item selection changes in the SSE branch currently.)

**Comment by adomasven** [comment]
We are. I played around with showing the selected item name, but if is too long then it looks awkward and if you cut it off with an ellipsis it looks even more awkward.
